src/ref.rs

```diff
@@ -616,6 +616,36 @@ where
         // SAFETY: We don't call any methods on `b` other than those provided by
         // `IntoByteSlice`.
         let b = unsafe { r.into_byte_slice() };
+        let b = b.into_byte_slice();
+
+        if let crate::layout::SizeInfo::Sized { .. } = T::LAYOUT.size_info {
+            let ptr = Ptr::from_ref(b);
+            // SAFETY: `T::raw_from_ptr_len` promises to preserve address and provenance in the referent
+            let ptr: Ptr<'_, T, _> = unsafe {
+                ptr.transmute_unchecked(|ptr| {
+                    let ptr = T::raw_from_ptr_len(
+                        ptr.as_non_null().cast(),
+                        <T::PointerMetadata as crate::PointerMetadata>::from_elem_count(0),
```

Comment:
Can you add a comment here explaining that this is redundant (since we've checked that `T: Sized` in this if branch's conditional), but we have to do it because the type system doesn't know that `T: Sized`?

Same in other functions.